### PR TITLE
FEAT Add SurvTRACE

### PR DESCRIPTION
**Pull request by Vincent-Maladiere**
This PR aims at refactoring and packaging the [SurvTRACE](https://github.com/RyanWangZf/SurvTRACE/tree/main) model. This effort focuses on:
- Generalizing the preprocessing and feature engineering of SurvTRACE so that we can run it on any dataset (instead of the three provided).
- Decoupling the preprocessing from the model, which relies on a `STConfig` global dictionary. We want to remove this global config and use hyper-parameters the scikit-learn way.
- Simplifying the codebase and removing some of the boilerplate by using skorch. Also, skorch will enable pipelining and cross-validation operations.

cc @ogrisel

### CODE REVIEW

File: hazardous/survtrace/_preprocessor.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+    def fit(self, y):
+        # Check strictly increasing
+        if not np.all(self.horizons[:-1] < self.horizons[1:]):
+            raise ValueError(
+                f"'horizon' must be strictly increasing, got {self.horizons=!r}"
+            )
+
+        # Defining cuts
+        mask_first_event = y["event"] == 1
```
Comment (Vincent-Maladiere) [inline hazardous/survtrace/_preprocessor.py:52]: This follows the SurvTRACE implementation, but we can probably use an "any event" mask instead.

File: hazardous/survtrace/_encoder.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
+import numpy as np
```
Comment (ogrisel) [inline hazardous/survtrace/_encoder.py:1]: Please add a header to note that this code is a refactoring derived from
https://github.com/RyanWangZf/SurvTRACE also published under the MIT license
with the following copyright:

```
Copyright (c) 2021 Maximum
```

**Comment by Vincent-Maladiere** [comment]
This a "working WIP", upon running fit I get similar performances for training and validation errors as in the original SurvTRACE version.

Next steps: get the final metrics and compare results.

```python
from hazardous.data._seer import (
    load_seer,
    CATEGORICAL_FEATURES,
    NUMERICAL_FEATURES,
)
from hazardous.survtrace._model import SurvTRACE


X, y = load_seer("hazardous/data/seer_cancer_cardio_raw_data.txt")
print(X.shape, y.shape)  # (476746, 28), (476746, 2)

model = SurvTRACE(
    numerical_features=NUMERICAL_FEATURES,
    categorical_features=CATEGORICAL_FEATURES,
)
model.fit(X, y)
```

cc @ogrisel, I think this is reviewable :)

**Comment by Vincent-Maladiere** [comment]
This PR is being split:
- the SEER part is being developed at https://github.com/soda-inria/hazardous/pull/24
- the SurvTRACE part is being developed in the branch `icml-2024`

**Comment by ogrisel** [comment]
Let's close this and later open a new PR dedicated to the model then.
